service/cecd: Implement some functions

NarcolepticK opened the pull request:
I stubbed functions I could find information on. There are a few more in the header than I actually stubbed.

EDIT: wwylele shared his old notes/code and I will see what I can manage to implement.

Updated behavior, should be more accurate now. Please try activating streetpass, it should write a number of files, and then go into system settings and deactivate streetpass to watch the folder disappear. Reactivating in game writes the saved streetpass info back into the system archive. As there is a limit to activated streetpasses, the archive is used for currently activated apps.

If needed navigate to your user folder, /nand/data/00000000000000000000000000000000/sysdata/ and delete the 00010026/ folder.

Increased debug logging info. Add Service.CECD:Debug to the logging to see what goes on.

[List of Streetpass Enabled Titles](https://gist.github.com/NarcolepticK/5da5f5b7d58bb0b2deace860b4d38aa9) (If there are any to add, let me know)

More to come.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4005)
<!-- Reviewable:end -->

FearlessTobi commented:
This PR fixes https://github.com/citra-emu/citra/issues/2475 due to the stub of `OpenAndRead`.

NarcolepticK commented:
Error codes found in the disassembly so far:

0xC82044B4  : FS  : Status : Nothing Happened : Exists Already
0xC8610BF3  :  CEC : Status : Out of Resource  :  Out of Memory
0xC8804464  :  FS : Status : Not Found : Not Found
0xC8804465  :  FS : Status : Not Found : Archive Not Mounted
0xC920181A  :  OS : Status : Canceled : Session Closed by Remote
0xC92044E6  :  FS :  Status : Canceled : Invalid Open Flags
0xC9604518  :  FS :  Status : Internal : Unacceptable
0xC8E1086A  :  CEC : Status : Invalid Argument : Invalid Signature
0xC8E1086B  :  CEC : Status : Invalid Argument : Not Found            
0xC8810BEA  :  CEC : Status : Not Found : Not Authorized
0xC8E10BE9  :  CEC : Status : Invalid Argument  : Too Large

0xD8601837  :  OS : Permanent : Out of Resource : Unknown (55)
0xD8604659  :  FS : Permanent : Out of Resource : Out of Resource
0xD8A0C7F8  :  Friends : Permanent : Invalid State : Not Initialized
0xD8A0C7F9  :  Friends :Permanent : Invalid State : Already Initialized
0xD900182F  : OS : Permanent : Wrong Argument : Invalid Command Header
0xD9004586  : FS : Permanent : Wrong Argument : Verification Failed

0xE0810BF8  :  CEC : Usage : Not Found : Not Initialized
0xE0E01BF2  :  OS : Usage : Invalid Argument : Misaligned Size
0xE0E046BC  :  FS : Usage : Invalid Argument : Invalid Argument
0xE0E0C7F6  :  Friends : Usage : Invalid Argument : Invalid Pointer  
0xE0E10BFD  :  CEC : Usage : Invalid Argument : Out of Range

Will update as I continue to find more.

NarcolepticK commented:
I haven't seen the CecStateAbbreviated enum in versions 9 through 11. When was this originally introduced?

I have found:
(UNKNOWN)
NDM_STATUS_WORKING
NDM_STATUS_IDLE
NDM_STATUS_SUSPENDING
NDM_STATUS_SUSPENDED

(UNKNOWN)
NONE
START
RESET_START
READYSCAN
READYSCANWAIT
STARTSCAN
RESCAN
NDM_RESUME
NDM_SUSPEND
NDM_SUSPEND_IMMEDIATE
STOPWAIT
STOP
STOP_FORCE
STOP_FORCE_WAIT
RESET_FILTER
DAEMON_STOP
DAEMON_START
EXIT
OVER_BOSS
OVER_BOSS_FORCE
OVER_BOSS_FORCE_WAIT

(UNKNOWN)
ERROR   
ERROR   

NONE
INIT
WIRELESS_PARAM_SETUP
WIRELESS_READY
WIRELESS_START_CONFIG
SCAN
SCANNING
CONNECT
CONNECTING
CONNECTED
CONNECT_TCP
CONNECTING_TCP
CONNECTED_TCP
NEGOTIATION
SEND_RECV_START
SEND_RECV_INIT
SEND_READY
SEND
RECEIVE_READY
RECEIVE
CONNECTION_FINISH_TCP
CONNECTION_FINISH
SEND_POST
RECEIVE_POST
FINISHING
FINISH
OVER_BOSS
IDLE

CEC_TEST_000
CEC_TEST_001
CEC_TEST_002
CEC_TEST_003
CEC_TEST_004
CEC_TEST_005
CEC_TEST_006

MerryMage commented:
@NarcolepticK Are you happy with having this in canary?

NarcolepticK commented:
@MerryMage I'd be alright with that. 

I need some more time to figure this out before I can really code more. What I have to test with is limited; I don't have Tomodachi Life for example. The more information and feedback, the better.

I don't feel I have hit a dead end yet. I have only just started getting into the disassembly. I am getting closer to having more to add.

FearlessTobi commented:
@NarcolepticK If you own Kid Icarus Uprising, that's another game that currently deadlocks with cecd commands on master. I'm unable to test it with your PR cause I'm on vacation.

NarcolepticK commented:
@FearlessTobi Thanks for the updates! I heard from Jhon591 that Kid Icarus Uprising works with this PR as well.

NarcolepticK commented:
I was able to find the initial state of the 00010026 CECD archive. It consists of eventlog.dat, /CEC, and /CEC/MBoxList____. The initial state of those two files were simple to write out and is documented in the code.

FearlessTobi commented:
@NarcolepticK No problem. Just as a side note, if you need more people for testing or want to coordinate with the other devs, be free to join either our Discord server or our citra-dev IRC channel ;D

NarcolepticK commented:
I saw beyblade burst crashed due to cecd in the forums, so I added a MBox_DIR check on failure to open file in OpenAndRead and OpenAndWrite and implemented CreateAndPopulateMBoxDirectory to the best of my current knowledge. I aim to make it more accurate, but the files and directories required will be created now if they don't exist.

NarcolepticK commented:
The InBox___/BoxInfo_____ and OutBox__/BoxInfo_____ have been updated to default values based on a fresh activation of mii plaza streetpass. They shouldn't throw invalid signature errors anymore...

JackoboLeChocobo commented:
Tested the code modifications with latest citra master code with Kid Icarus : boot the first time, but not the second time ( freezes in the loading startup screen ).

NarcolepticK commented:
@JackoboLeChocobo Are you able to navigate to your user folder, /nand/data/00000000000000000000000000000000/sysdata/00010026/00000000/CEC/ and delete the <8 digit id> folder for Kid Icarus Uprising. I think it's 00030000. And then try again? Currently the modifications write the data when the mbox <id> folder needs to be created.

JackoboLeChocobo commented:
@NarcolepticK So I will re-apply the code and I will retest that with your advice. I will try to do that quickly.

JackoboLeChocobo commented:
Congrats ! You win a lot of cookies ! :cookie: :cookie: :cookie: :cookie: :cookie: :cookie: :cookie: 

I look after nand/data/00000000000000000000000000000000/sysdata/00010026/00000000/CEC/ . And I see indeed a directory called 00030000. After I delete it ( before launching or during the launch of the game ), it doesn't hang anymore \o/

NarcolepticK commented:
@JackoboLeChocobo That's good to hear. I'm currently looking into forming mboxinfo____ correctly. I have found that the title_id that needs to be written doesn't seems to contain the region related info. The mboxdata.050 contains more specific title_id related info and may very well not directly match the folder <id>.

JackoboLeChocobo commented:
@NarcolepticK

Do you need a new test on the game with new commit f9b87bf ?

NarcolepticK commented:
@JackoboLeChocobo  If you can, that would be helpful. Thanks!

JackoboLeChocobo commented:
@NarcolepticK 

Can reboot without issue ^^, but now I have the texts that don't dissapear when the characters talk

JackoboLeChocobo commented:
@JackoboLeChocobo  I will retest it so : I work on a RetroArch improved version of Citra so it can be my build fault. Time for the chocobo to retest it ^^

JackoboLeChocobo commented:
@NarcolepticK , After testing, it's not an issue with your code, but an another issue that we don't have before because Kid Icarus never started before your fix code ^^ . We need more people to confirm the subtitles issues. To test that : 
- 1°) Have a clean "citra-emu" user clean
- 2°) Start the game
- 3°) Create a new file on the game
- 4°) Select "Solo" mode
- 5°) Listen to Palutena's beautiful voice on gameplay/tutorial lesson ( you will get subtitles during the explanation )
- 6°) Confirm that these subtitles display correctly or not

FearlessTobi commented:
![unbenannt](https://user-images.githubusercontent.com/20753089/44308972-3a7aaf80-a3bf-11e8-9264-e73d46afe5c4.PNG)

I'm able to reproduce this problem. An issue for it should be made **after** this PR is merged.

JackoboLeChocobo commented:
@FearlessTobi and @NarcolepticK

:cake: :cake: :cake: :cake: :cake: :cake: Congrats ! You win a lot of cakes !  :cake: :cake: :cake: :cake: :cake: :cake:

"Accurate multiplication" enabled, the game nows has no subtitle issues !
Thank you all for the tests and works ! If you need new tests, don't hesitate to ask me if I can do it.

JackoboLeChocobo commented:
@NarcolepticK I applied changes on my RetroArch build. Kid Icarus still boots without issue ( tested two reboots with a clean save dir at the beginning )

NarcolepticK commented:
Silly clang-format fixup... missed the header include section, I run clang-format -i before I upload and it caught the other parts...

NarcolepticK commented:
Rebased and corrected eventlog.dat creation.

wwylele commented on src/core/hle/service/cecd/cecd.h:
```diff
@@ -4,47 +4,481 @@
+        u16_le padding2;
+        u16_le num_boxes; /// 24 max
+        u16_le padding3;
+        u8 box_names[16 * 24]; /// 16 char names, 24 boxes
```
`std::array<std::array<u16_le, 16>, 24> box_names`

wwylele commented on src/core/hle/service/cecd/cecd.h:
```diff
@@ -4,47 +4,481 @@
+        u64_le sender_id;
+        u64_le sender_id2;
+        struct Time {
+            u32_le a, b, c;
```
Would this be in the same format as `CecMBoxInfoHeader::Time`?

wwylele commented on src/core/hle/service/cecd/cecd.h:
```diff
@@ -4,47 +4,481 @@
+
+    struct CecMBoxInfoHeader {
+        u16_le magic; // 0x6363 'cc'
+        u16_le padding;
```
You can use `INSERT_PADDING_BYTES` from common/common_funcs.h for all the padding stuff

wwylele commented on src/core/hle/service/cecd/cecd.h:
```diff
@@ -18,28 +19,28 @@ class Module final {
-        CEC_COMMAND_OVER_BOSS_FORCE = 0x13,
-        CEC_COMMAND_OVER_BOSS_FORCE_WAIT = 0x14,
-        CEC_COMMAND_END = 0x15,
+        NONE = 0,
```
:stuck_out_tongue: I meant `PascalCase` like `None`, `Start` and `ResetStart`. This is our convention for enum item names

Dragios commented:
Is this still stub or proper implementation? If it is the latter please rename accordingly.

Dragios commented on src/core/hle/service/cecd/cecd_s.cpp:
```diff
@@ -12,22 +12,27 @@ CECD_S::CECD_S(std::shared_ptr<Module> cecd)
     static const FunctionInfo functions[] = {
         // cecd:u shared commands
```
`// cecd:s shared commands`

wwylele replied:
@Dragios I guess this comment means "these functions are shared with cecd:u"

NarcolepticK replied:
Cecd:s contains a set of shared commands with cecd:u, as well as its own set.

Dragios replied:
It seems I misread the comment. This can be ignored then. My apologize.

Dragios commented on src/core/hle/service/cecd/cecd_s.cpp:
```diff
@@ -12,22 +12,27 @@ CECD_S::CECD_S(std::shared_ptr<Module> cecd)
+        {0x00120104, &CECD_S::OpenAndRead, "OpenAndRead"},
+        {0x001E0082, nullptr, "GetEventLog"},
+        {0x001F0000, nullptr, "GetEventLogStart"},
+        {0x40020002, nullptr, "GetCecInfoEventHandleSys"},
```
Are you sure this is `0x40020002` and not `0x04020002` instead?
Correct me if I'm wrong because I refer here
![3dbrew cecd:s](https://user-images.githubusercontent.com/18496868/45353553-b1395000-b5ed-11e8-936d-5b48b9f0157e.png)

NarcolepticK replied:
@Dragios Thank you for noticing that.

wwylele commented:
Is this ready to merge? If no more comments I am going to merge tomorrow
